apps/mark-scan/pat-device-input/src/pat_input_reader.rs

```diff
@@ -134,83 +146,125 @@ impl fmt::Display for MockPin {
+        let mut probe_error = None;
+        if address > POSSIBLE_PIN_OFFSET {
+            probe_error = Some(io::Error::new(
+                io::ErrorKind::AddrNotAvailable,
+                "Test error",
+            ))
+        }
+
+        MockNoOffsetPin {
+            address,
+            active: false,
+            probe_error,
+        }
+    }
+
+    fn probe(&self) -> Result<(), io::Error> {
+        if let Some(probe_error) = &self.probe_error {
+            // io::Error doesn't implement clone()
+            return Err(io::Error::new(probe_error.kind(), probe_error.to_string()));
+        }
+        Ok(())
+    }
+
+    fn set_up(&self) -> Result<(), io::Error> {
+        Ok(())
+    }
+
+    fn tear_down(&self) {}
+
+    fn is_active(&self) -> bool {
+        self.active
+    }
+}
+
+impl fmt::Display for MockNoOffsetPin {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.address)
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use vx_logging::set_app_name;
 
     use super::*;
 
     fn get_default_mocked_reader() -> PatInputReader<MockPin> {
-        get_mocked_reader(None)
-    }
-
-    fn get_mocked_reader(connection_error: Option<io::Error>) -> PatInputReader<MockPin> {
-        PatInputReader {
-            is_connected_pin: Some(MockPin {
-                address: IS_CONNECTED_PIN_ADDRESS,
-                active: false,
-                probe_error: probe_err,
-            }),
-            signal_a_pin: Some(MockPin {
-                address: SIGNAL_A_PIN_ADDRESS,
-                active: false,
-                probe_error: None,
-            }),
-            signal_b_pin: Some(MockPin {
-                address: SIGNAL_B_PIN_ADDRESS,
-                active: false,
-                probe_error: None,
-            }),
-        }
+        let mut reader: PatInputReader<MockPin> = PatInputReader::new();
+        let result = reader.connect();
+        assert!(result.is_ok());
+        reader
     }
 
     #[test]
     fn test_connect_offset() {
         set_app_name("test");
-        let mut reader = get_default_mocked_reader();
-        reader.connect().unwrap();
+        let reader = get_default_mocked_reader();
+
         assert_eq!(
             reader.is_connected_pin.unwrap().address,
             IS_CONNECTED_PIN_ADDRESS + POSSIBLE_PIN_OFFSET
         );
+        assert_eq!(
+            reader.signal_a_pin.unwrap().address,
+            SIGNAL_A_PIN_ADDRESS + POSSIBLE_PIN_OFFSET
+        );
+        assert_eq!(
+            reader.signal_b_pin.unwrap().address,
+            SIGNAL_B_PIN_ADDRESS + POSSIBLE_PIN_OFFSET
+        );
     }
 
     #[test]
     fn test_connect_no_offset() {
         set_app_name("test");
-        let mut reader =
-            get_mocked_reader(Some(io::Error::new(io::ErrorKind::NotFound, "Test Error")));
-        reader.connect().unwrap();
+        let mut reader: PatInputReader<MockNoOffsetPin> = PatInputReader::new();
```

Comment:
Not sure which version you find more readable, but you could avoid repeating the struct name:
```suggestion
        let mut reader = PatInputReader::<MockNoOffsetPin>::new();
```